File: examples/basic_usage.rs
//! HAZE API Usage Examples

use haze::{KeyPair, Tokenomics, FogEconomy, MistbornAsset, DensityLevel};
use haze::types::sha256;
use std::collections::HashMap;

fn main() {
    println!("HAZE Blockchain - Usage Examples\n");

    // Example 1: Creating a key pair
    println!("=== Example 1: Creating a Key Pair ===");
    let keypair = KeyPair::generate();
    let address = keypair.address();
    println!("Address: {}", hex::encode(address));
    println!();

    // Example 2: Working with tokenomics
    println!("=== Example 2: HAZE Tokenomics ===");
    let tokenomics = Tokenomics::new();
    println!("Initial supply: {} HAZE", tokenomics.total_supply());
    println!("Current inflation: {}%", tokenomics.inflation_rate() as f64 / 100.0);
    
    // Staking
    let validator = address;
    tokenomics.stake(validator, validator, 1_000_000_000_000_000_000).unwrap(); // 1000 HAZE
    println!("Staked: 1000 HAZE");
    
    if let Some(stake) = tokenomics.get_stake(&validator) {
        println!("Stake information: {} HAZE", stake.amount);
    }
    println!();

    // Example 3: Creating a Mistborn NFT
    println!("=== Example 3: Creating Mistborn Asset ===");
    let asset_id = sha256(b"example_asset");
    let mut asset = MistbornAsset::create(
        asset_id,
        address,
        DensityLevel::Ethereal,
        HashMap::from([
            ("name".to_string(), "Legendary Sword".to_string()),
            ("rarity".to_string(), "legendary".to_string()),
        ]),
    );
    println!("NFT created: {}", hex::encode(&asset_id[..8]));
    println!("Density level: Ethereal (5KB)");
    
    // Condensing NFT (increasing density)
    let mut new_data = HashMap::new();
    new_data.insert("texture".to_string(), "sword_texture.png".to_string());
    new_data.insert("model".to_string(), "sword_model.gltf".to_string());
    asset.condense(new_data, None).unwrap(); // None = no blob storage needed for Light density
    println!("NFT condensed: Light (50KB)");
    println!();

    // Example 4: Fog Economics
    println!("=== Example 4: Fog Economics ===");
    let economy = FogEconomy::new();
    
    // Update game activity
    economy.update_game_activity(
        "game_1".to_string(),
        1_000_000_000,
        address,
    ).unwrap();
    println!("Game activity updated for game_1");
    
    // Create liquidity pool
    let pool_id = economy.create_liquidity_pool(
        "HAZE".to_string(),
        "GOLD".to_string(),
        1_000_000_000_000_000_000, // 1M HAZE
        10_000_000_000_000_000_000, // 10M GOLD
        30, // 0.3% fee
    ).unwrap();
    println!("Liquidity pool created: {}", pool_id);
    
    // Create Vortex Market
    let market_id = economy.create_vortex_market(
        "game_1".to_string(),
        vec![
            ("HAZE".to_string(), "GOLD".to_string()),
            ("HAZE".to_string(), "SILVER".to_string()),
        ],
        haze::MarketConditions::ArbitrageOpportunity { discount: 5 },
        24, // 24 hours
    ).unwrap();
    println!("Vortex Market created: {}", market_id);
    println!();

    println!("All examples executed successfully!");
}

File: sdk/examples/basic-usage.ts
/**
 * Basic usage examples for HAZE SDK
 */

import {
  HazeClient,
  KeyPair,
  TransactionBuilder,
  MistbornAsset,
  FogEconomy,
  DensityLevel,
  DEFAULT_API_URL,
} from '../src/index';

async function main() {
  console.log('HAZE Blockchain SDK - Basic Usage Examples\n');

  // Initialize client
  const client = new HazeClient({
    baseUrl: process.env.HAZE_API_URL || DEFAULT_API_URL,
  });

  try {
    // Example 1: Health check
    console.log('=== Example 1: Health Check ===');
    const health = await client.healthCheck();
    console.log('Health:', health);
    console.log();

    // Example 2: Get blockchain info
    console.log('=== Example 2: Blockchain Info ===');
    const info = await client.getBlockchainInfo();
    console.log('Current height:', info.current_height);
    console.log('Total supply:', info.total_supply.toString(), 'HAZE');
    console.log('Current wave:', info.current_wave);
    console.log();

    // Example 3: Generate key pair
    console.log('=== Example 3: Generate Key Pair ===');
    const keyPair = await KeyPair.generate();
    const address = keyPair.getAddressHex();
    console.log('Address:', address);
    console.log('Public key:', keyPair.getPublicKeyHex());
    console.log();

    // Example 4: Get account balance
    console.log('=== Example 4: Get Account Balance ===');
    try {
      const balance = await client.getBalance(address);
      console.log('Balance:', balance.toString(), 'HAZE');
    } catch (error: any) {
      console.log('Account not found or error:', error.message);
    }
    console.log();

    // Example 5: Create and sign a transfer transaction
    console.log('=== Example 5: Create Transfer Transaction ===');
    const recipient = await KeyPair.generate();
    const recipientAddress = recipient.getAddress();
    
    const transferTx = TransactionBuilder.createTransfer(
      keyPair.getAddress(),
      recipientAddress,
      BigInt(1000000000000000000), // 1000 HAZE (with 18 decimals)
      BigInt(1000000000000000),    // 0.001 HAZE fee
      0 // nonce
    );

    const signedTx = await TransactionBuilder.sign(transferTx, keyPair);
    const txHash = TransactionBuilder.getHashHex(signedTx);
    console.log('Transaction hash:', txHash);
    console.log('Transaction type:', signedTx.type);
    console.log('Amount:', signedTx.amount.toString(), 'HAZE');
    console.log();

    // Example 6: Create Mistborn Asset
    console.log('=== Example 6: Create Mistborn Asset ===');
    const assetId = MistbornAsset.createAssetId('legendary_sword_001');
    const assetTx = MistbornAsset.createCreateTransaction(
      assetId,
      keyPair.getAddress(),
      DensityLevel.Ethereal,
      {
        name: 'Legendary Sword',
        rarity: 'legendary',
        game_id: 'fantasy_rpg',
      },
      [
        { name: 'attack', value: '100', rarity: 0.95 },
        { name: 'defense', value: '50', rarity: 0.85 },
      ],
      'fantasy_rpg'
    );

    const signedAssetTx = await MistbornAsset.sign(assetTx, keyPair);
    console.log('Asset ID:', MistbornAsset.assetIdToHex(assetId));
    console.log('Density:', signedAssetTx.data.density);
    console.log('Metadata:', signedAssetTx.data.metadata);
    console.log();

    // Example 7: Fog Economy - Get liquidity pools
    console.log('=== Example 7: Fog Economy ===');
    const economy = new FogEconomy(client);
    try {
      const pools = await economy.getLiquidityPools();
      console.log('Liquidity pools:', pools.length);
      if (pools.length > 0) {
        const pool = pools[0];
        console.log('Pool ID:', pool.pool_id);
        console.log('Assets:', pool.asset1, '/', pool.asset2);
        console.log('Reserves:', pool.reserve1.toString(), '/', pool.reserve2.toString());
      }
    } catch (error: any) {
      console.log('No liquidity pools found or error:', error.message);
    }
    console.log();

    // Example 8: Condense asset (increase density)
    console.log('=== Example 8: Condense Asset ===');
    const condenseTx = MistbornAsset.createCondenseTransaction(
      assetId,
      keyPair.getAddress(),
      DensityLevel.Light,
      {
        texture: 'sword_texture.png',
        model: 'sword_model.gltf',
      },
      [
        { name: 'glow_effect', value: 'true', rarity: 0.9 },
      ]
    );
    const signedCondenseTx = await MistbornAsset.sign(condenseTx, keyPair);
    console.log('Condensed to:', signedCondenseTx.data.density);
    console.log('New metadata:', signedCondenseTx.data.metadata);
    console.log();

    console.log('All examples completed successfully!');
  } catch (error: any) {
    console.error('Error:', error.message);
    if (error.stack) {
      console.error('Stack:', error.stack);
    }
    process.exit(1);
  }
}

// Run examples
if (require.main === module) {
  main().catch(console.error);
}
